### PR TITLE
CombinatorialRangeAttribute

### PR DESCRIPTION
**Pull request by mille5a9**
This is concerning Isssue #18 

This new CombinatorialRangeAttribute can be used to test with a range of integers. The attribute will count up from a number if given two integers, or count by "x" from a starting value to an ending value if given three integers. My aim is to satisfy some of the functionality of Nunit's [Range] attribute, as CombinatorialValuesAttribute already covers Nunit's [Values] attribute and I sometimes need to convert projects from Nunit to Xunit in my work.

This is my first attempt at an open-source contribution, please let me know if I've missed anything.

### CODE REVIEW

File: src/Xunit.Combinatorial/CombinatorialRangeAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            if (to < from)
+            {
+                throw new ArgumentOutOfRangeException(nameof(to));
+            }
+
+            // step must be a positive integer
+            if (step < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(step));
+            }
```
Comment (AArnott) [inline src/Xunit.Combinatorial/CombinatorialRangeAttribute.cs]: While these are reasonable constraints, they aren't documented. Can you add to the doc comments for these parameters stating that `from <= to` and `step >= 1`?
Also please document what happens if `from == to`, as well as what happens if `step` jumps *over* `to` rather than stepping to a value equal to it.

File: src/Xunit.Combinatorial.Tests/CombinatorialRangeAttributeTests.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+namespace Xunit.Combinatorial.Tests
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+
+    public class CombinatorialRangeAttributeTests
```
Comment (AArnott) [inline src/Xunit.Combinatorial.Tests/CombinatorialRangeAttributeTests.cs]: These are good tests. Can you add tests that actually *use* `CombinatorialRangeAttribute`? While the test body itself may not be able to *assert* much (aside from the arg being _one of_ the expected values), it would be good to see it demonstrated and verify visually that Test Explorer or the test log actually invoked the test that many times.

Comment (AArnott) [reply src/Xunit.Combinatorial.Tests/CombinatorialRangeAttributeTests.cs]: hint: I think you'll find the new attribute doesn't work as-is. I think you'll need to change `ValuesUtilities.cs` to recognize the added attribute type.

Comment (mille5a9) [reply src/Xunit.Combinatorial.Tests/CombinatorialRangeAttributeTests.cs]: Added another local variable to `GetValuesFor()` to cover both Combinatorial attributes. I recognize that this solution is not good for adding more attributes in the future but a better course of action did not come to me easily, so I hope this is okay.

The test explorer is showing all 10 cases from the test I added to `SampleUses.cs` and they're passing with some bare-bones assertions.

File: src/Xunit.Combinatorial/CombinatorialRangeAttribute.cs (hunk 1 above)
Comment (mille5a9) [reply src/Xunit.Combinatorial/CombinatorialRangeAttribute.cs]: Docs have been amended. I added explanations of the edge cases in the xml and also added some basic information / a sample test to the readme.

File: src/Xunit.Combinatorial/ValuesUtilities.cs [hunk 3]
```diff
@@ -21,6 +21,13 @@ internal static IEnumerable<object> GetValuesFor(ParameterInfo parameter)
             Requires.NotNull(parameter, nameof(parameter));
 
             var valuesAttribute = parameter.GetCustomAttribute<CombinatorialValuesAttribute>();
+            var rangeAttribute = parameter.GetCustomAttribute<CombinatorialRangeAttribute>();
+
+            if (rangeAttribute != null)
+            {
+                return rangeAttribute.Values;
+            }
+
```
Comment (AArnott) [inline src/Xunit.Combinatorial/ValuesUtilities.cs]: This is good, but you inserted it in between the existing local and its use. Can you move this to *after* the use of `valuesAttribute`?

Comment (mille5a9) [reply src/Xunit.Combinatorial/ValuesUtilities.cs]: Rearranged the lines, thanks for your help!

File: src/Xunit.Combinatorial.Tests/SampleUses.cs [hunk 4]
```diff
@@ -58,5 +76,23 @@ public CustomValuesAttribute()
...
+        }
+
+        [AttributeUsage(AttributeTargets.Parameter)]
+        private class CustomStepRangeAttribute : CombinatorialRangeAttribute
+        {
+            public CustomStepRangeAttribute()
+                : base(0, 3, 2)
+            {
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mille5a9) [reply src/Xunit.Combinatorial.Tests/SampleUses.cs]: I was taking after the custom derived attribute for `CombinatorialValues`, although I don't know why it wasn't directly used in the first place.

I have gone ahead and used CombinatorialRange in the test.

**Comment by AArnott** [comment]
Thanks!
